### ISSUE TITLE
Error during the correction step: unable to find 'corjob.files'

### ISSUE THREAD

**Issue by gringer** [marbl/canu#224]
My assembly failed today due to not being able to find corjob.files:

```
-- Finished on Wed Aug 10 17:51:17 2016 (4851 seconds) with 474.9 GB free disk space
----------------------------------------
-- generate corrected reads attempt 1 begins with 14 finished, and 1 to compute.
----------------------------------------
-- Starting concurrent execution on Wed Aug 10 17:51:17 2016 with 474.9 GB free disk space (1 processes; 6 concurrently)

    /home/gringer/bioinf/MIMR-2016-Jul-01-GBIS/GLG/reads/reads_Jodie_NbAL5_gDNA_2016_Jul_08/downloads/canu_NippoAllSensitive_auto/correction/2-correction/correctReads.sh 5 > /home/gringer/bioinf/MIMR-2016-Jul-01-GBIS/GLG/reads/reads_Jodie_NbAL5_gDNA_2016_Jul_08/downloads/canu_NippoAllSensitive_auto/correction/2-correction/correctReads.000005.out 2>&1

-- Finished on Wed Aug 10 18:27:56 2016 (2199 seconds) with 473.2 GB free disk space
----------------------------------------
--
-- 1 read correction jobs failed:
--   job /home/gringer/bioinf/MIMR-2016-Jul-01-GBIS/GLG/reads/reads_Jodie_NbAL5_gDNA_2016_Jul_08/downloads/canu_NippoAllSensitive_auto/correction/2-correction/correction_outputs/0005.fasta FAILED.
-- generate corrected reads attempt 2 begins with 14 finished, and 1 to compute.
----------------------------------------
-- Starting concurrent execution on Wed Aug 10 18:27:56 2016 with 473.2 GB free disk space (1 processes; 6 concurrently)

    /home/gringer/bioinf/MIMR-2016-Jul-01-GBIS/GLG/reads/reads_Jodie_NbAL5_gDNA_2016_Jul_08/downloads/canu_NippoAllSensitive_auto/correction/2-correction/correctReads.sh 5 > /home/gringer/bioinf/MIMR-2016-Jul-01-GBIS/GLG/reads/reads_Jodie_NbAL5_gDNA_2016_Jul_08/downloads/canu_NippoAllSensitive_auto/correction/2-correction/correctReads.000005.out 2>&1

-- Finished on Wed Aug 10 18:54:57 2016 (1621 seconds) with 473.2 GB free disk space
----------------------------------------
-- Concatenating correctReads output.
================================================================================
Don't panic, but a mostly harmless error occurred and canu failed.

canu failed with 'can't open '/home/gringer/bioinf/MIMR-2016-Jul-01-GBIS/GLG/reads/reads_Jodie_NbAL5_gDNA_2016_Jul_08/downloads/canu_NippoAllSensitive_auto/correction/2-correction/corjob.files' for reading: No such file or directory'.

```

**Comment by gringer** [comment]
I notice that corjob.files is only accessed at three points in the code:

```
$ grep -r -B 3 -A 3 'corjob.files' canu-1.3
canu-1.3/Linux-amd64/bin/lib/canu/CorrectReads.pm-  finishStage:
canu-1.3/Linux-amd64/bin/lib/canu/CorrectReads.pm-    print STDERR "-- Found ", scalar(@successJobs), " read correction output files.\n";
canu-1.3/Linux-amd64/bin/lib/canu/CorrectReads.pm-
canu-1.3/Linux-amd64/bin/lib/canu/CorrectReads.pm:    open(L, "> $path/corjob.files") or caExit("failed to open '$path/corjob.files'", undef);
canu-1.3/Linux-amd64/bin/lib/canu/CorrectReads.pm-    print L @successJobs;
canu-1.3/Linux-amd64/bin/lib/canu/CorrectReads.pm-    close(L);
canu-1.3/Linux-amd64/bin/lib/canu/CorrectReads.pm-
--
canu-1.3/Linux-amd64/bin/lib/canu/CorrectReads.pm-    my $files = 0;
canu-1.3/Linux-amd64/bin/lib/canu/CorrectReads.pm-    my $reads = 0;
canu-1.3/Linux-amd64/bin/lib/canu/CorrectReads.pm-
canu-1.3/Linux-amd64/bin/lib/canu/CorrectReads.pm:    open(F, "< $path/corjob.files")                           or caExit("can't open '$path/corjob.files' for reading: $!", undef);
canu-1.3/Linux-amd64/bin/lib/canu/CorrectReads.pm-    open(N, "< $wrk/$asm.gkpStore/readNames.txt")             or caExit("can't open '$wrk/$asm.gkpStore/readNames.txt' for reading: $!", undef);
canu-1.3/Linux-amd64/bin/lib/canu/CorrectReads.pm-    open(O, "| gzip -1c > $WRK/$asm.correctedReads.fasta.gz") or caExit("can't open '$WRK/$asm.correctedReads.fasta.gz' for writing: $!", undef);
canu-1.3/Linux-amd64/bin/lib/canu/CorrectReads.pm-    #open(O, "> $WRK/$asm.correctedReads.fasta")               or caExit("can't open '$WRK/$asm.correctedReads.fasta' for writing: $!", undef);
--
canu-1.3/Linux-amd64/bin/lib/canu/CorrectReads.pm-    my $Nlog     = 0;
canu-1.3/Linux-amd64/bin/lib/canu/CorrectReads.pm-
canu-1.3/Linux-amd64/bin/lib/canu/CorrectReads.pm-    if (getGlobal("saveReadCorrections") != 1) {
canu-1.3/Linux-amd64/bin/lib/canu/CorrectReads.pm:        open(F, "< $path/corjob.files") or caExit("can't open '$path/corjob.files' for reading: $!", undef);
canu-1.3/Linux-amd64/bin/lib/canu/CorrectReads.pm-        while (<F>) {
canu-1.3/Linux-amd64/bin/lib/canu/CorrectReads.pm-            chomp;
canu-1.3/Linux-amd64/bin/lib/canu/CorrectReads.pm-
```

**Comment by brianwalenz** [comment]
Check for errors in correction/2-correet ction/correctReads.000005.out.

This looks like that job failed (at least twice) yet canu keep running as if it succeeded.

**Comment by gringer** [comment]
Yep, seems to have been an error in 0005; listing of correction directory:

```
$ ls canu_NippoAllSensitive_auto/correction/2-correction/correction_outputs/
0001.dump.success  0002.err           0003.fasta         0005.err            0006.fasta         0008.dump.success  0009.err           0010.fasta         0012.dump.success  0013.err           0014.fasta
0001.err           0002.fasta         0004.dump.success  0005.fasta.WORKING  0007.dump.success  0008.err           0009.fasta         0011.dump.success  0012.err           0013.fasta         0015.dump.success
0001.fasta         0003.dump.success  0004.err           0006.dump.success   0007.err           0008.fasta         0010.dump.success  0011.err           0012.fasta         0014.dump.success  0015.err
0002.dump.success  0003.err           0004.fasta         0006.err            0007.fasta         0009.dump.success  0010.err           0011.fasta         0013.dump.success  0014.err           0015.fasta

```

**Comment by gringer** [comment]
Segfault fun:

```
$ cat canu_NippoAllSensitive_auto/correction/2-correction/correction_outputs/0005.err

Failed with 'Segmentation fault'

Backtrace (mangled):

/home/gringer/install/canu/canu-1.3/Linux-amd64/bin/falcon_sense[0x40723d]
/lib/x86_64-linux-gnu/libpthread.so.0(+0x10ed0)[0x7f44e1d8ded0]
/home/gringer/install/canu/canu-1.3/Linux-amd64/bin/falcon_sense[0x40a53f]
/home/gringer/install/canu/canu-1.3/Linux-amd64/bin/falcon_sense[0x4078e0]
/home/gringer/install/canu/canu-1.3/Linux-amd64/bin/falcon_sense[0x4092d7]
/home/gringer/install/canu/canu-1.3/Linux-amd64/bin/falcon_sense[0x40257a]
/lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xf0)[0x7f44e19fc730]
/home/gringer/install/canu/canu-1.3/Linux-amd64/bin/falcon_sense[0x4032f9]

Backtrace (demangled):

[0] /home/gringer/install/canu/canu-1.3/Linux-amd64/bin/falcon_sense() [0x40723d]
[1] /lib/x86_64-linux-gnu/libpthread.so.0::(null) + 0x10ed0  [0x7f44e1d8ded0]
[2] /home/gringer/install/canu/canu-1.3/Linux-amd64/bin/falcon_sense() [0x40a53f]
[3] /home/gringer/install/canu/canu-1.3/Linux-amd64/bin/falcon_sense() [0x4078e0]
[4] /home/gringer/install/canu/canu-1.3/Linux-amd64/bin/falcon_sense() [0x4092d7]
[5] /home/gringer/install/canu/canu-1.3/Linux-amd64/bin/falcon_sense() [0x40257a]
[6] /lib/x86_64-linux-gnu/libc.so.6::(null) + 0xf0  [0x7f44e19fc730]
[7] /home/gringer/install/canu/canu-1.3/Linux-amd64/bin/falcon_sense() [0x4032f9]

GDB:


```

**Comment by brianwalenz** [comment]
Yay.

Can you build with debug (gmake BUILDDEBUG=1) and rerun.  This _should_ populate the demangled backtrace.

Would you be willing to share the output of generateCorrectionLayouts for debugging?

With a little effort, you can narrow down the -b/-e range to find the read that is failing, and share only that input.  Of course, once you do find the read that fails, you split this job into two pieces, run each piece, and join the output to make it look like the job finished successfully.

**Comment by gringer** [comment]
I suspect it might be caused by a really long read (e.g. over 200kb). I have generateCorrectionLayout runtimes of less than a minute for ~250 reads when it doesn't segfault, and over 10 minutes when it does.

**Comment by gringer** [comment]
Okay, so I've attached the single read generateCorrectionLayout that appears to be causing the issue.
[correctionLayout_falcon_sense_fail.txt.gz](https://github.com/marbl/canu/files/412547/correctionLayout_falcon_sense_fail.txt.gz)

**Comment by gringer** [comment]
Removing reads longer than 50 kbp (which for now appear to only be base-called noise) allowed Canu to assemble my dataset.

**Comment by brianwalenz** [comment]
Thanks for tracking down and sharing the read.  I was able to reproduce a crash after _six_ _hours_.  I don't think read length is the problem here, rather, the low complexity sequence giving the aligner a fit.

**Comment by skoren** [comment]
I tested the fix for issue #318 and the read processes without error so this will be fixed when issue #318 is resolved.
